### ISSUE TITLE
\includegraphicsをGUI化

### ISSUE THREAD

**Issue by Yasunari15** [doraTeX/TeX2img#92]
TeXShopではWindowに画像をDrag & Dropすると`\includegraphics[]{hope.pdf}`のように補完されるようですが，同じ機能がTeX2imgにもあると良いかと思います。ただ「Sourceを復元」と同じUIになるため，実装して下さる場合には，何らかの処置が必要かと思われます。どう思われますでしょうか。

**Comment by doraTeX** [comment]
おっしゃるように「ソースを復元」との重複が防げないので、困難かと思います。

**Comment by Yasunari15** [comment]
「Sourceを復元」は「既存のFileを読み込む」と似ているため，Drag & Dropへの対応を

* 「Input TeX code」では`\includegraphics[]{hope.pdf}`
* 「Input TeX source」では「Sourceを復元」

としてみたら解決するのではないでしょうか。

**Comment by doraTeX** [comment]
よくわからないのですが、毎回ダイアログを出して選択を促すということですか？

**Comment by Yasunari15** [comment]
![screen shot 2018-07-20 at 14 25 41](https://user-images.githubusercontent.com/38091102/42984748-128ecd9c-8c29-11e8-8d0e-b0401ab112dc.png)

**Comment by doraTeX** [comment]
それは直観的ではないですね。ソース復元する際は上のエリアに復元されるわけですから。

**Comment by Yasunari15** [comment]
では「Input TeX source or PDF/PS/EPS file」の上に画像をDragすると，薄めの文字で`\includegraphics`が現れて，それを「Input TeX Code」にDropするというのは，どうでしょうか。強引すぎる気もしますが。

**Comment by doraTeX** [comment]
そんなUIはmacOSアプリで見たことがないですね……
ありうるとしたら、環境設定パネルに「ドラッグ&ドロップの効果」みたいな選択項目を設けるくらいでしょうか……

**Comment by Yasunari15** [comment]
XcodeのObjectsからDesignし始めるときは，右上に当てるようにしてから，現れたWindowを配置していくようなものだった覚えがあります。

**Comment by doraTeX** [comment]
> XcodeのObjectsからDesignし始めるときは，右上に当てるようにしてから，現れたWindowを配置していくようなもの

それは自分は見たことがないですね……

**Comment by doraTeX** [comment]
簡単な方法としては，

* オプションを押しながらドラッグしたら `\includegraphics` を挿入する

という方法が考えられます。オプションキーというのは機能を切り替えるためのキーなので，その動きはUIとしてそれなりに直感的と言っていいのではないかという気もします。

ただし，次なる懸念として，挿入されるものを `\includegraphics` ではなくいろいろとカスタマイズしたい，という要望が出てきたときに，挿入文字列のカスタマイズ機能を設けないとするのか，それともそれをカスタマイズするUIをどこかに何らかの形で設けるか，と考え始めると，なかなかやっかいです。
（ちなみに，TeXShopではドラッグ＆ドロップで自動挿入される文字列をカスタマイズする機能は提供していません。）

**Comment by Yasunari15** [comment]
> オプションを押しながらドラッグしたら \includegraphics を挿入する

macOSらしい直感的な操作で素晴らしいと思います。

> 挿入文字列のカスタマイズ機能を設けないとするのか，それともそれをカスタマイズするUIをどこかに何らかの形で設けるか

`\includegraphics`以外の何かを思いつきませんでしたので，`\includegraphics`ということにして，要望があれば応じて対策を考える，ということになさったらいかがでしょうか。TeX2epsには書式に合わせた`\img`があると良いかもしれませんね。

**Comment by doraTeX** [comment]
TeXShopでは，

* `.tex` をドラッグ＆ドロップしたら `\input{/path/to/hoge.tex}` を挿入
* `.pdf`をドラッグ＆ドロップしたら `\includegraphics{/path/to/hoge.pdf}` を挿入 
……

のように拡張子ごとに挿入コードが固定的に規定されており，それらのカスタマイズ法は用意されていません。

**Comment by Yasunari15** [comment]
TeX2imgでは，

* `.tex`は`\input{/path/to/hoge.tex}`を挿入
* 画像は`Sourceの復元`を試みる
* Optionがあれば`\includegraphics{/path/to/hoge.png}`のように挿入

とすれば良いのでは無いでしょうか。TeXShopへの要望として，

* `.styは`\usepackage{/path/to/hoge}`を挿入

があるとPackage製作のときに便利かと思います。まずは，TeX2imgに取り入れて下さると嬉しいです。

**Comment by doraTeX** [comment]
> .texは\input{/path/to/hoge.tex}を挿入

現状は `.tex` のドラッグ＆ドロップも「ファイルの読み込み」ですので，`\input{/path/to/hoge.tex}` の挿入もも option ありの場合に限定すべきです。

> TeXShopへの要望として，
> .styは `\usepackage{/path/to/hoge}`を挿入

それは既にあるはずです。

**Comment by Yasunari15** [comment]
> \input{/path/to/hoge.tex} の挿入も option ありの場合に限定すべきです。

そうですね。この場合に，PreambleをDropした`.tex`に合わせて上書きしているようですが，QuitしたときにDefaultに戻しておいた方が良いように思えます。

Preambleを編集した際も，現状の`Manage templates`や`Save current as a template`があるのでDefaultにしても良い気がしますが。

**Comment by doraTeX** [comment]
> PreambleをDropした.texに合わせて上書きしているようですが，QuitしたときにDefaultに戻しておいた方が良いように思えます。

> Defaultにしても良い気がしますが。

それは設計思想に反します。そもそも，「Default とは何か」というのも規定されていませんし，TeX2imgは「前回終了時のプリアンブルを保持する代わりに終了時に「破棄してよろしいですか」という

**Comment by Yasunari15** [comment]
なるほど。Preambleの編集を反映したい際は，`Save current as a template`から同じ名前で保存してReplaceを選択することにします。

では，FileをDrag & Dropする挙動はそのままに，OptionをつけるとTeXShopのように「拡張子ごとに挿入コードを補完」するようにしたらどうでしょうか。

**Comment by doraTeX** [comment]
> では，FileをDrag & Dropする挙動はそのままに，OptionをつけるとTeXShopのように「拡張子ごとに挿入コードを補完」するようにしたらどうでしょうか。

今のところ，optionをつけてドラッグ＆ドロップするとTeXShopと同じコードを補完する設計で考えています。補完コードはTeXShopと同一に仕様をそろえておくのがよいでしょうね。

**Comment by Yasunari15** [comment]
よろしくお願い致します。実装の難易度は分かりませんし，お忙しいとは思いますが「#87」「#89」も合わせて2.3.1となると嬉しいです。

**Comment by Yasunari15** [comment]
> それは既にあるはずです。

試してみたところ`\usepackage{/path/to/hoge}`ではなく，ただ単に`\usepackage{hoge}`と補完されるようです。意図的にDrag & Dropしている場合には`\usepackage{/path/to/hoge}`とした方が良いと思います。
